src/basic/in-addr-util.c

```diff
@@ -226,6 +226,70 @@ int in_addr_prefix_next(int family, union in_addr_union *u, unsigned prefixlen)
         return -EAFNOSUPPORT;
 }
 
+int in_addr_prefix_nth(int family, union in_addr_union *u, unsigned prefixlen, unsigned nth) {
+        assert(u);
+
+        if (prefixlen <= 0)
+                return 0;
+
+        if (nth == 0) { // identity??
+                return 1;
+        }
+
+        if (family == AF_INET) {
+                uint32_t c, n;
+                if (prefixlen > 32)
+                        prefixlen = 32;
+
+                c = be32toh(u->in.s_addr);
+
+                n = c + (nth * (1UL << (32 - prefixlen)));
+
+                if (n < c)
+                        return 0; // overflow
```

Comment:
hmm these kind of overflow checks are a bit sloppy, since according to the C standard the result of oveflowed + and * isn't really defined, even though on most CPUs this just runs over. i.e please check this properly before doing the addition

Reply:
I think this should now cover all of the cases properly.